logstash-core/lib/logstash/plugins/builtin/pipeline/input.rb

```diff
@@ -68,9 +68,11 @@ def internalReceive(events)
         stream_position = stream_position + 1
       end)
       ReceiveResponse.completed()
-    rescue java.lang.InterruptedException, IOError => e
-      # maybe an IOException in enqueueing
-      logger.debug? && logger.debug('queueing event failed', message: e.message, exception: e.class, backtrace: e.backtrace)
+    rescue java.lang.InterruptedException, org.logstash.ackedqueue.QueueRuntimeException, java.io.IOException, IOError => e
+      # maybe an IOException in enqueueing, majority of IOExceptions are not wrapped with Ruby IOError
+      # WARN is intentional, if ensure_delivery disabled we lose warnings in PipelineBus.java caused by ReceiveResponse.failed_at(stream_position, e)
+      logger.warn('queueing event failed', message: e.message, exception: e.class)
+      logger.debug? && logger.debug('queueing event failed trace', message: e.message, exception: e.class, backtrace: e.backtrace)
```

Comment:
A couple things here:

1) `@queue << event` routes through `JrubyAbstractQueueWriteClient#rubyPush`, which in turn routes to `JRubyAbstractQueueWriteClient#doPush`, whose PQ implementation routes through `JRubyAckedQueueExt#rubyWrite` which catches all `java.io.IOException` and re-casts them as ruby `IOError`s (and whose MQ implementation doesn't throw `IOException` at all). Therefore we don't need to add `java.io.IOException` to the rescue clause.
2) The pipeline bus is capable of logging for itself, including making decisions based on the value of its `ensureDelivery`. I would recommend _keeping_ this log message at debug-level and _adding_ warn-on-drop message to the pipeline bus.

Reply:
1. Yup, confused `@queue.class` when changing and we do have a single entry which throws `IOError`, removed `java.io.IOException`.
This might be out of scope now but I think retrying sending process on `IOError` also doesn't make sense to me. Consequently failure point relates to IO operation which are unhandleable in an application level.

Reply:
2. We could place it in `PipelineBus.java#sendEvents()` (as shown example below) but any thread accessing this point makes noise even after volatile input becomes unavailable in the address bus. - confirmed to be sure.

```
...
    if (!ensureDelivery && !sendWasSuccess) {
        logger.warn("`ensure_delivery` is disabled in pipeline-to-pipeline communication. Some event sends failed, throwing remaining inflight {} events.", orderedEvents.length - lastFailedPosition);
    }
} while (partialProcessing);
```

<img width="1476" alt="image" src="https://github.com/elastic/logstash/assets/99575341/23a34e2b-f53e-4688-8772-1f99aae019c3">